src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -68,6 +74,34 @@ public void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
                 source.Cancel();
             }
+
+            ClearDiagnostics(workspace);
+
+            AssertCleanup();
+        }
+
+        private void ClearDiagnostics(Workspace workspace)
+        {
+            var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
+            var args = _service.GetDiagnosticsUpdatedEventArgs(workspace, projectId: null, documentId: null, cancellationToken: CancellationToken.None);
```

Comment:
I'm really struggling to understand what these two lines of code do.  We're calling in to 'Clear Diagnostics'.  But the way we clear diagnostics is to first create an incremental analyzer... then ask it to update diagnostics arguments...  then go through those arguments and tell the analyzer to remove documents from it?

Basically, this code is *really* confusing.  I can't figure out what it's really trying to do, or why it solves the problem.  For example: why is it only removing documents and not removing projects?  How did these documents get added in the first place?  

Why are we asking the "service" for diagnostic args.. but then telling the analyzer to remove documents?

If it is the analyzer that is holding onto data for too long, why don't we just ask the analyzer itself to remove any data it should not be holding onto?  That seems like it would make a lot more sense.  This approach means that this type here has to know precisely how the analyzer is caching data, and what precisely it should call to clear it up.

Basically it seems like in order to understand this code you need to understand deeply how many complex pieces end up holding onto data and how to make them stop doing that.  however, i don't know how anyone would know to fix this place up if, for example, the analyzer held onto any more state in the future.